vendor/bat-native-ledger/src/bat/ledger/internal/wallet/wallet_util.cc

```diff
@@ -0,0 +1,189 @@
+
+namespace ledger {
+namespace wallet {
+
+namespace {
+
+std::string WalletTypeToState(const std::string& wallet_type) {
+  if (wallet_type == constant::kWalletBitflyer) {
+    return state::kWalletBitflyer;
+  }
+
+  if (wallet_type == constant::kWalletGemini) {
+    return state::kWalletGemini;
+  }
+
+  if (wallet_type == constant::kWalletUphold) {
+    return state::kWalletUphold;
+  }
+
+  NOTREACHED();
+  return "";
+}
+
+} //namespace
+
+type::ExternalWalletPtr GetWallet(LedgerImpl* ledger, const std::string wallet_type) {
+  DCHECK(ledger);
+  const std::string wallet_string =
+      ledger->ledger_client()->GetEncryptedStringState(WalletTypeToState(wallet_type));
+
+  if (wallet_string.empty()) {
+    return nullptr;
+  }
+
+  absl::optional<base::Value> value = base::JSONReader::Read(wallet_string);
+  if (!value || !value->is_dict()) {
+    BLOG(0, "Parsing of " + wallet_type +" wallet failed");
+    return nullptr;
+  }
+
+  base::DictionaryValue* dictionary = nullptr;
+  if (!value->GetAsDictionary(&dictionary)) {
+    BLOG(0, "Parsing of " + wallet_type + " wallet failed");
+    return nullptr;
+  }
+
+  auto wallet = ledger::type::ExternalWallet::New();
+  wallet->type = wallet_type;
+
+  auto* token = dictionary->FindStringKey("token");
+  if (token) {
+    wallet->token = *token;
+  }
+
+  auto* address = dictionary->FindStringKey("address");
+  if (address) {
+    wallet->address = *address;
+  }
+
+  auto* one_time_string = dictionary->FindStringKey("one_time_string");
+  if (one_time_string) {
+    wallet->one_time_string = *one_time_string;
+  }
+
+  auto status = dictionary->FindIntKey("status");
+  if (status) {
+    wallet->status = static_cast<ledger::type::WalletStatus>(*status);
+  }
+
+  auto* user_name = dictionary->FindStringKey("user_name");
+  if (user_name) {
+    wallet->user_name = *user_name;
+  }
+
+  auto* verify_url = dictionary->FindStringKey("verify_url");
+  if (verify_url) {
+    wallet->verify_url = *verify_url;
+  }
+
+  auto* add_url = dictionary->FindStringKey("add_url");
+  if (add_url) {
+    wallet->add_url = *add_url;
+  }
+
+  auto* withdraw_url = dictionary->FindStringKey("withdraw_url");
+  if (withdraw_url) {
+    wallet->withdraw_url = *withdraw_url;
+  }
+
+  auto* account_url = dictionary->FindStringKey("account_url");
+  if (account_url) {
+    wallet->account_url = *account_url;
+  }
+
+  auto* login_url = dictionary->FindStringKey("login_url");
+  if (login_url) {
+    wallet->login_url = *login_url;
+  }
+
+  auto* fees = dictionary->FindDictKey("fees");
+  if (fees) {
+    base::DictionaryValue* fees_dictionary;
+    if (fees->GetAsDictionary(&fees_dictionary)) {
+      for (base::DictionaryValue::Iterator it(*fees_dictionary);
+          !it.IsAtEnd();
+          it.Advance()) {
+        if (!it.value().is_double()) {
+          continue;
+        }
+
+        wallet->fees.insert(std::make_pair(it.key(), it.value().GetDouble()));
+      }
+    }
+  }
+
+  return wallet;
```

Comment:
how do we know this is a valid wallet? All fields are optional so this could be completely empty